packages/frontend/src/components/rich_text_editor/commands.ts

```diff
@@ -0,0 +1,120 @@
+import type { NodeType } from "prosemirror-model";
+import { type Command, type EditorState, NodeSelection, type Transaction } from "prosemirror-state";
+import { hasContent } from "./utils";
+
+export function insertMathDisplayCmd(nodeType: NodeType, initialText = ""): Command {
+    return (state: EditorState, dispatch: ((tr: Transaction) => void) | undefined) => {
+        const { $from, $to } = state.selection;
+
+        if (!dispatch) {
+            return true;
+        }
+
+        // There is an idiomatic pattern at the start of commands inserting new nodes to check that the node can
+        // be inserted with canReplaceWith. In this case we skip the check because canReplaceWith will
+        // always (I think) fail when trying to put a block inside a paragraph.
+
+        const selectedText = state.doc.textBetween($from.pos, $to.pos, " ");
+        const initialTextContent = initialText || selectedText;
+        const initialContent = initialTextContent ? state.schema.text(initialTextContent) : null;
+
+        const mathNode = nodeType.create({}, initialContent);
+
+        let tr = state.tr;
+
+        // delete the selected text (it will be replaced with `initalContent`)
+        if ($from.pos !== $to.pos) {
+            tr = tr.delete($from.pos, $to.pos);
+        }
+
+        // if we're inside a paragraph, split the paragraph
+        if ($from.parent.type.name === "paragraph" && $from.parent.content.size !== 0) {
+            tr = tr.split($from.pos);
+        }
+
+        tr = tr.insert($from.pos, mathNode);
+        tr = tr.setSelection(NodeSelection.create(tr.doc, $from.pos + 1));
+
+        dispatch(tr);
+
+        return true;
+    };
+}
+
+// Currently does not work due to bug in the automerge-prosemirror plugin. (It also might not work in
+// general, but it theoretically should)
+export function insertMathInlineCmd(mathNodeType: NodeType, initialText = ""): Command {
```

Comment:
Inline math is at least as important as display math. How do we debug this problem? Is the next step to ask the Automerge devs for help?

Also, if I'm not mistaken, this function comes from here: https://github.com/benrbray/prosemirror-math/blob/master/lib/commands/insert-math-cmd.ts

We should include a link to the upstream definition.